### PR TITLE
Feat/FSRS-5

### PR DESCRIPTION
**Pull request by L-M-Sherlock**

### CODE REVIEW

**Review by AlexErrant — COMMENTED** [review]
https://github.com/open-spaced-repetition/fsrs-rs/pull/197/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542

Just wondering; should we synchronize versions across FSRS? Like, could this be `5.0.0`?

**Comment by L-M-Sherlock** [comment]
> Just wondering; should we synchronize versions across FSRS? Like, could this be `5.0.0`?

There are too many packages of FSRS implementation, and some of them are not maintained by myself. So I think it's unnecessary to synchronize versions across FSRS.

File: src/optimal_retention.rs [hunk 1]
```diff
@@ -215,20 +247,24 @@ pub fn simulate(
...
         // Sample 'rating' for 'need_review' entries
         let mut ratings = Array1::zeros(deck_size);
-        izip!(&mut ratings, &(&need_review & !&forget))
-            .filter(|(_, &condition)| condition)
-            .for_each(|(rating, _)| {
-                *rating = review_rating_choices[review_rating_dist.sample(&mut rng)]
+        izip!(&mut ratings, &need_review, &forget)
+            .filter(|(_, &condition, _)| condition)
+            .for_each(|(rating, _, forget)| {
+                if *forget {
```
Comment (asukaminato0721) [inline src/optimal_retention.rs:253]:
```rs
*y = if x { t1 } else { t2 }
```
?

**Comment by user1823** [comment]
@L-M-Sherlock, can you summarize the main changes in FSRS 5 as compared to FSRS 4.5? I believe that it will be useful for the reviewers of this PR as well as for future reference.

**Comment by L-M-Sherlock** [comment]
> can you summarize the changes in FSRS 5 as compared to FSRS 4.5

In short, it considers the short-term schedule, and changes the formula to calculate the initial diffculty.

For details, see https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#fsrs-5-draft

**Comment by L-M-Sherlock** [comment]
@AlexErrant https://github.com/open-spaced-repetition/fsrs-rs/actions/runs/9928079913/job/27423861989 The merge action failed. Could you help me check it?

**Comment by AlexErrant** [comment]
Done!
